pubtools/pulplib/_impl/client/client.py

```diff
@@ -179,6 +181,36 @@ def search_repository(self, criteria=None):
             response_f, lambda data: self._handle_page(Repository, search, data)
         )
 
+    def _do_upload_file(self, upload_id, repo_id, filename):
+        tasks_f = f_return([])
+        offset = 0
+
+        def _do_next_part_upload(accumulated_tasks, upload_data):
+            data, upload_id, offset = upload_data
+            upload_task_f = self._do_upload(data, upload_id, offset)
+
+            return f_map(
+                upload_task_f, lambda upload_task: accumulated_tasks + upload_task
+            )
+
+        # calculate size and checksum of the file during upload
+        with open(filename, "rb") as f:
+            checksum = hashlib.sha256()
+            size = 0
+            data = f.read(self._CHUNK_SIZE)
+            LOG.info("uploading %s to repo %s", filename, repo_id)
+            while data:
```

Comment:
This is getting closer, but I think it's still doing too much in blocking style.  Specifically, it's reading entire file into memory before returning.

I don't think it will scale properly when there's many uploads.  A RHEL release for example may need to upload 40,000 files.  With the current approach here, if someone loops through all the files and calls upload for each, it will cause *all* content of *all* the files to be loaded in the memory of the current process, which probably won't work.

I'd expect that, to be scalable, the reading & checksumming of the file also has to be done in non-blocking style, i.e. you only read the next chunk once the current chunk has finished uploading.

Now that being said, I understand your current use-case isn't really needing to be scalable, so it's not a blocker to fix this - as long as the API isn't affected.  e.g. if we can make it more scalable later without changing the API, then it's acceptable to leave it like this (and probably file an issue to make it scalable).